src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/StaticInvalidationService.kt

```diff
@@ -0,0 +1,156 @@
+import com.intellij.refactoring.suggested.endOffset
+import com.intellij.refactoring.suggested.startOffset
+
+class StaticInvalidationService {
+
+    // HashMap of Class filePath -> (HashMap of Method Signature -> (Method Body, Covered Lines))
+    private var savedMethods: HashMap<String, HashMap<String, Pair<ArrayList<PsiElement>, Set<Int>>>> = HashMap()
+
+    /**
+     * Checks if method has been changed since last test generation (no, if first run)
+     * @param signature the method in question
+     * @param body list of the body-elements of the method, without whitespaces
+     * @param methods hashmap of previously tested methods and their bodies
+     * @return the lines which should be deleted (based on previous lines for method)
+     */
+    private fun validateMethod(
+        signature: String,
+        body: Pair<ArrayList<PsiElement>, Set<Int>>,
+        methods: HashMap<String, Pair<ArrayList<PsiElement>, Set<Int>>>
+    ): Set<Int> {
+        val savedBody = methods[signature]
+
+        // if body doesn't exist, method seen first time
+        // if amount of elements in body different, method surely changed
+        if (savedBody == null || body.first.size != savedBody.first.size) {
+            methods[signature] = body
+            return setOf()
+        }
+        // compare each element (no whitespace)
+        body.first.zip(savedBody.first).forEach {
+            if (!it.first.text.equals(it.second.text)) {
+                methods[signature] = body
+                return savedBody.second
+            }
+        }
+        methods[signature] = Pair(savedBody.first, body.second)
+        return setOf()
+    }
+
+    /**
+     * Checks for class what methods within it have been modified
+     * @param filePath path where the class is located
+     * @param methods the methods of the class and the lines they cover
+     * @param className the name of the class (used for preciser hashing)
+     * @return the lines in the class that should be deleted (based on previous lines for method)
+     */
+    private fun validateClass(
+        filePath: String,
+        methods: HashMap<String, Pair<ArrayList<PsiElement>, Set<Int>>>,
+        className: String
+    ): MutableSet<Int> {
+        val linesToDiscard = mutableSetOf<Int>()
+
+        // get old methods
+        val methodsSaved = savedMethods.getOrPut("$filePath/$className") { methods }
+        // validate each method against old methods
+        methods.keys.forEach {
+            val changed = validateMethod(it, methods[it]!!, methodsSaved)
+            linesToDiscard.addAll(changed)
+        }
+
+        return linesToDiscard
+    }
+
+    /**
+     * Method to invalidate the changed parts of a cache.
+     *
+     * @param fileUrl the url of a file
+     * @param lines the lines to invalidate tests
+     * @param cache the cache
+     */
+    fun invalidateCacheLines(fileUrl: String, lines: Set<Int>, cache: TestCaseCachingService) {
```

Comment:
Maybe it would make more sense to inject the cache service into the static invalidation service instead of passing it as an argument (i.e. project.service<...>`)

Reply:
@kirilvasilev16's domain

Reply:
I decided to look at it and came to the conclusion the service can be accessed easily through the project. Changes have been made. Point should be addressed now.